### PR TITLE
Multi account and SSO

### PR DESCRIPTION
**Pull request by stefan-niedermann**
grafik

# Open issues

## Caching (ETag, Modified)

- [x] Move `e tag` and `last modified` into `account`-table in database
- [x] Clean up old `e tag` and `last modified` SharedPreferences in database upgrade process
- [x] :construction: Something is wrong with the modified date - all my notes appear as last modified 1970.

## Migration

- [x] In database upgrade process create first account entry in `account`-table with the existing entry from the SharedPreferences (also copy `etag` and `last modified` from SharedPreferences to first account entry)
- [x] Remove the now no longer needed SharedPreferences for the account settings
- [x] Make sure that this not yet authenticated account asks smoothly the SSO for permission on first sync
  - [x] Android 9
  - [ ] :construction: Android 8 (to be tested)
  - [x] :construction: Android 7
  - [ ] :construction: Android 6 (to be tested)
  - [ ] :construction: Android 5 (to be tested)
  - [x] :construction: Android 4
- [x] Indexes for the `table_accounts`
- [ ] :construction: Foreign key from `account_id` in `table_notes` to `id` in `table_accounts` does not work yet
- [x] Remove TrustPreference from PreferencesFragment

## Widgets

- [x] @dan0xii the widgets do currently not store the account id from the note. They will have to, otherwise will switching account break the widgets.
- [x] Widgets do not switch the active account in the app
- [x] Note list widget appears to only show one account regardless of which is active in the main app
- [x] Single note widget says "Note not found" after upgrade

## Dark theme

- [x] The font in the header and the new accounts list in the drawer look ugly and wrong with the dark theme.

# Follow-Up issues

- [ ] Move widget configurations from SharedPreferences to database

### CODE REVIEW

**Comment by stefan-niedermann** [comment]
I think this works pretty good already.

- [ ] @korelstar  I think i messed something up with `e tag` and `last modified` - It's always `null` respectively `0` which causes much unnecessary data traffic and the sync takes ages. Might you have a look please?

grafik

- [ ] @dan0xii The font in the header and the new accounts list in the drawer look ugly and wrong with the dark theme. Maybe you can do the same dark magic as you did for the categories?

grafik

- [ ] Make accounts deletable
- [ ] Anyone around who know something about databases? I think we might need some indexes for the `table_accounts` and maybe a foreign key from `account_id` in `table_notes` to `id` in `table_accounts`? I am just unsure if the foreign key breaks the migration, when we add `1` as `account_id` while no single-sign-on-account exists yet.

**Comment by dan0xii** [comment]
>  @dan0xii the widgets do currently not store the account id from the note. They will have to, otherwise will switching account break the widgets.

Is the intention that if the user opens a note from the home screen via the single note widget and that note belongs to an account that isn't currently active, the app should switch accounts?

Are you having the same results as I am (API 27) when adding an account: the login process works as expected but then the Notes app becomes the Files app and there doesn't appear to be a way to go "back" to the now-authenticated Notes app other than closing it from android's recent apps list and re-opening.

**Comment by stefan-niedermann** [comment]
> Is the intention that if the user opens a note from the home screen via the single note widget and that note belongs to an account that isn't currently active, the app should switch accounts?

Yes, opening a note from single note widget of a different account should switch the account.

> Are you having the same results as I am (API 27) when adding an account: the login process works as expected but then the Notes app becomes the Files app and there doesn't appear to be a way to go "back" to the now-authenticated Notes app other than closing it from android's recent apps list and re-opening.

According to the video on the Github page of SSO it seems to be the workflow when adding a new account from a third party app which has not been configured on the files app yet. If you are picking an already existing account this does not happen. Though i will have a talk with the SSO developers to change this behavior. I agree that adding account from third party app should return the user to the third party app.

**Comment by stefan-niedermann** [comment]
@dan0xii can you make sure to migrate the already existing SharedPreferences in the onUpgrade-Method in SQLiteOpenHelper? Just set the correspondenting accountId hard coded to `1`.

File: app/src/main/java/it/niedermann/owncloud/notes/android/activity/NotesListViewActivity.java [hunk 1]
```diff
@@ -233,6 +220,27 @@ protected void onCreate(Bundle savedInstanceState) {
 
     @Override
     protected void onResume() {
+        // TODO This needs to skip if there are no detected changes between localAccount id
+        // and current active single sign on account
+        try {
+            selectAccount(SingleAccountHelper.getCurrentSingleSignOnAccount(getApplicationContext()).name);
+        } catch (NoCurrentAccountSelectedException | NextcloudFilesAppAccountNotFoundException e) {
+            handleNotAuthorizedAccount();
```
Comment (stefan-niedermann) [inline app/src/main/java/it/niedermann/owncloud/notes/android/activity/NotesListViewActivity.java]: @dan0xii this breaks the migration of an account from a previous version:

`handleNotAuthorizedAccount()` pops up a dialog to import/authorize this account, after pressing `cancel` or `ok` the activity is resumed which leads to some kind of endless-loop.

Comment (stefan-niedermann) [reply app/src/main/java/it/niedermann/owncloud/notes/android/activity/NotesListViewActivity.java]: Maybe we should call this in `onResume()` **and** in `onCreate()` - and call `handleNotAuthorizedAccount()` only in `onCreate()` while removing the exception handling in `onResume()`...?

Or a boolean flag where we remember if we have asked the user already to import/authorize the account?

Comment (stefan-niedermann) [reply app/src/main/java/it/niedermann/owncloud/notes/android/activity/NotesListViewActivity.java]: I went for a boolean flag for now to simply fix this.

**Comment by dan0xii** [comment]
> Single note widget says "Note not found" after upgrade

@stefan-niedermann, How do I recreate this?

**Comment by stefan-niedermann** [comment]
Install v 1.0.0, create single note widget of any note. Update to the SSO branch and no it says "No note found" instead of the content. I can provide a screen record if you have troubles reproducing it?

**Comment by dan0xii** [comment]
I'm hitting a problem where the upgrade from v1.0.0 to the latest commit in this branch gives me the error "Something went wrong.. please try again" and I'm dropped at a blank notes list. The drawer shows my nextcloud account (the same account v1.0.0 was configured with), I tap on it and I see the account selection dialog flicker and then disappear before the same error message is displayed.

I thought it was something I was doing but seems to happen without my latest changes. Have you had it happen to you?

**Comment by dan0xii** [comment]
It appears to be happening when there's no files client installed. I hadn't realised I'd reset the VM and hadn't re-installed it. Will upload a video if needed.

Logcat:
```
2019-10-16 17:54:37.176 5945-5945/? I/art: Not late-enabling -Xcheck:jni (already on)
2019-10-16 17:54:37.176 5945-5945/? W/art: Unexpected CPU variant for X86 using defaults: x86_64
2019-10-16 17:54:37.213 5945-5945/it.niedermann.owncloud.notes W/System: ClassLoader referenced unknown path: /data/app/it.niedermann.owncloud.notes-2/lib/x86_64
2019-10-16 17:54:37.232 5945-5945/it.niedermann.owncloud.notes W/art: Before Android 4.1, method android.graphics.PorterDuffColorFilter androidx.vectordrawable.graphics.drawable.VectorDrawableCompat.updateTintFilter(android.graphics.PorterDuffColorFilter, android.content.res.ColorStateList, android.graphics.PorterDuff$Mode) would have incorrectly overridden the package-private method in android.graphics.drawable.Drawable
2019-10-16 17:54:37.269 5945-5945/it.niedermann.owncloud.notes I/art: Rejecting re-init on previously-failed class java.lang.Class<androidx.core.view.ViewCompat$2>: java.lang.NoClassDefFoundError: Failed resolution of: Landroid/view/View$OnUnhandledKeyEventListener;
2019-10-16 17:54:37.269 5945-5945/it.niedermann.owncloud.notes I/art:     at void androidx.core.view.ViewCompat.setOnApplyWindowInsetsListener(android.view.View, androidx.core.view.OnApplyWindowInsetsListener) (ViewCompat.java:2421)
2019-10-16 17:54:37.269 5945-5945/it.niedermann.owncloud.notes I/art:     at android.view.ViewGroup androidx.appcompat.app.AppCompatDelegateImpl.createSubDecor() (AppCompatDelegateImpl.java:779)
2019-10-16 17:54:37.269 5945-5945/it.niedermann.owncloud.notes I/art:     at void androidx.appcompat.app.AppCompatDelegateImpl.ensureSubDecor() (AppCompatDelegateImpl.java:659)
2019-10-16 17:54:37.269 5945-5945/it.niedermann.owncloud.notes I/art:     at void androidx.appcompat.app.AppCompatDelegateImpl.setContentView(int) (AppCompatDelegateImpl.java:552)
2019-10-16 17:54:37.269 5945-5945/it.niedermann.owncloud.notes I/art:     at void androidx.appcompat.app.AppCompatActivity.setContentView(int) (AppCompatActivity.java:161)
2019-10-16 17:54:37.269 5945-5945/it.niedermann.owncloud.notes I/art:     at void it.niedermann.owncloud.notes.android.activity.NotesListViewActivity.onCreate(android.os.Bundle) (NotesListViewActivity.java:185)
2019-10-16 17:54:37.269 5945-5945/it.niedermann.owncloud.notes I/art:     at void android.app.Activity.performCreate(android.os.Bundle) (Activity.java:6679)
2019-10-16 17:54:37.269 5945-5945/it.niedermann.owncloud.notes I/art:     at void android.app.Instrumentation.callActivityOnCreate(android.app.Activity, android.os.Bundle) (Instrumentation.java:1118)
2019-10-16 17:54:37.269 5945-5945/it.niedermann.owncloud.notes I/art:     at android.app.Activity android.app.ActivityThread.performLaunchActivity(android.app.ActivityThread$ActivityClientRecord, android.content.Intent) (ActivityThread.java:2618)
2019-10-16 17:54:37.269 5945-5945/it.niedermann.owncloud.notes I/art:     at void android.app.ActivityThread.handleLaunchActivity(android.app.ActivityThread$ActivityClientRecord, android.content.Intent, java.lang.String) (ActivityThread.java:2726)
2019-10-16 17:54:37.269 5945-5945/it.niedermann.owncloud.notes I/art:     at void android.app.ActivityThread.-wrap12(android.app.ActivityThread, android.app.ActivityThread$ActivityClientRecord, android.content.Intent, java.lang.String) (ActivityThread.java:-1)
2019-10-16 17:54:37.269 5945-5945/it.niedermann.owncloud.notes I/art:     at void android.app.ActivityThread$H.handleMessage(android.os.Message) (ActivityThread.java:1477)
2019-10-16 17:54:37.269 5945-5945/it.niedermann.owncloud.notes I/art:     at void android.os.Handler.dispatchMessage(android.os.Message) (Handler.java:102)
2019-10-16 17:54:37.269 5945-5945/it.niedermann.owncloud.notes I/art:     at void android.os.Looper.loop() (Looper.java:154)
2019-10-16 17:54:37.269 5945-5945/it.niedermann.owncloud.notes I/art:     at void android.app.ActivityThread.main(java.lang.String[]) (ActivityThread.java:6119)
2019-10-16 17:54:37.269 5945-5945/it.niedermann.owncloud.notes I/art:     at java.lang.Object java.lang.reflect.Method.invoke!(java.lang.Object, java.lang.Object[]) (Method.java:-2)
2019-10-16 17:54:37.269 5945-5945/it.niedermann.owncloud.notes I/art:     at void com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run() (ZygoteInit.java:886)
2019-10-16 17:54:37.269 5945-5945/it.niedermann.owncloud.notes I/art:     at void com.android.internal.os.ZygoteInit.main(java.lang.String[]) (ZygoteInit.java:776)
2019-10-16 17:54:37.269 5945-5945/it.niedermann.owncloud.notes I/art: Caused by: java.lang.ClassNotFoundException: Didn't find class "android.view.View$OnUnhandledKeyEventListener" on path: DexPathList[[zip file "/data/app/it.niedermann.owncloud.notes-2/base.apk"],nativeLibraryDirectories=[/data/app/it.niedermann.owncloud.notes-2/lib/x86_64, /system/lib64, /vendor/lib64]]
2019-10-16 17:54:37.269 5945-5945/it.niedermann.owncloud.notes I/art:     at java.lang.Class dalvik.system.BaseDexClassLoader.findClass(java.lang.String) (BaseDexClassLoader.java:56)
2019-10-16 17:54:37.269 5945-5945/it.niedermann.owncloud.notes I/art:     at java.lang.Class java.lang.ClassLoader.loadClass(java.lang.String, boolean) (ClassLoader.java:380)
2019-10-16 17:54:37.269 5945-5945/it.niedermann.owncloud.notes I/art:     at java.lang.Class java.lang.ClassLoader.loadClass(java.lang.String) (ClassLoader.java:312)
2019-10-16 17:54:37.269 5945-5945/it.niedermann.owncloud.notes I/art:     at void androidx.core.view.ViewCompat.setOnApplyWindowInsetsListener(android.view.View, androidx.core.view.OnApplyWindowInsetsListener) (ViewCompat.java:2421)
2019-10-16 17:54:37.269 5945-5945/it.niedermann.owncloud.notes I/art:     at android.view.ViewGroup androidx.appcompat.app.AppCompatDelegateImpl.createSubDecor() (AppCompatDelegateImpl.java:779)
2019-10-16 17:54:37.269 5945-5945/it.niedermann.owncloud.notes I/art:     at void androidx.appcompat.app.AppCompatDelegateImpl.ensureSubDecor() (AppCompatDelegateImpl.java:659)
2019-10-16 17:54:37.269 5945-5945/it.niedermann.owncloud.notes I/art:     at void androidx.appcompat.app.AppCompatDelegateImpl.setContentView(int) (AppCompatDelegateImpl.java:552)
2019-10-16 17:54:37.269 5945-5945/it.niedermann.owncloud.notes I/art:     at void androidx.appcompat.app.AppCompatActivity.setContentView(int) (AppCompatActivity.java:161)
2019-10-16 17:54:37.269 5945-5945/it.niedermann.owncloud.notes I/art:     at void it.niedermann.owncloud.notes.android.activity.NotesListViewActivity.onCreate(android.os.Bundle) (NotesListViewActivity.java:185)
2019-10-16 17:54:37.269 5945-5945/it.niedermann.owncloud.notes I/art:     at void android.app.Activity.performCreate(android.os.Bundle) (Activity.java:6679)
2019-10-16 17:54:37.269 5945-5945/it.niedermann.owncloud.notes I/art:     at void android.app.Instrumentation.callActivityOnCreate(android.app.Activity, android.os.Bundle) (Instrumentation.java:1118)
2019-10-16 17:54:37.269 5945-5945/it.niedermann.owncloud.notes I/art:     at android.app.Activity android.app.ActivityThread.performLaunchActivity(android.app.ActivityThread$ActivityClientRecord, android.content.Intent) (ActivityThread.java:2618)
2019-10-16 17:54:37.269 5945-5945/it.niedermann.owncloud.notes I/art:     at void android.app.ActivityThread.handleLaunchActivity(android.app.ActivityThread$ActivityClientRecord, android.content.Intent, java.lang.String) (ActivityThread.java:2726)
2019-10-16 17:54:37.269 5945-5945/it.niedermann.owncloud.notes I/art:     at void android.app.ActivityThread.-wrap12(android.app.ActivityThread, android.app.ActivityThread$ActivityClientRecord, android.content.Intent, java.lang.String) (ActivityThread.java:-1)
2019-10-16 17:54:37.269 5945-5945/it.niedermann.owncloud.notes I/art:     at void android.app.ActivityThread$H.handleMessage(android.os.Message) (ActivityThread.java:1477)
2019-10-16 17:54:37.269 5945-5945/it.niedermann.owncloud.notes I/art:     at void android.os.Handler.dispatchMessage(android.os.Message) (Handler.java:102)
2019-10-16 17:54:37.269 5945-5945/it.niedermann.owncloud.notes I/art:     at void android.os.Looper.loop() (Looper.java:154)
2019-10-16 17:54:37.269 5945-5945/it.niedermann.owncloud.notes I/art:     at void android.app.ActivityThread.main(java.lang.String[]) (ActivityThread.java:6119)
2019-10-16 17:54:37.269 5945-5945/it.niedermann.owncloud.notes I/art:     at java.lang.Object java.lang.reflect.Method.invoke!(java.lang.Object, java.lang.Object[]) (Method.java:-2)
2019-10-16 17:54:37.269 5945-5945/it.niedermann.owncloud.notes I/art:     at void com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run() (ZygoteInit.java:886)
2019-10-16 17:54:37.269 5945-5945/it.niedermann.owncloud.notes I/art:     at void com.android.internal.os.ZygoteInit.main(java.lang.String[]) (ZygoteInit.java:776)
2019-10-16 17:54:37.287 5945-5945/it.niedermann.owncloud.notes I/ViewConfigCompat: Could not find method getScaledScrollFactor() on ViewConfiguration
2019-10-16 17:54:37.396 5945-5945/it.niedermann.owncloud.notes W/System.err: com.nextcloud.android.sso.exceptions.NoCurrentAccountSelectedException: No account selected yet
2019-10-16 17:54:37.396 5945-5945/it.niedermann.owncloud.notes W/System.err:     at com.nextcloud.android.sso.helper.SingleAccountHelper.getCurrentAccountName(SingleAccountHelper.java:46)
2019-10-16 17:54:37.396 5945-5945/it.niedermann.owncloud.notes W/System.err:     at com.nextcloud.android.sso.helper.SingleAccountHelper.getCurrentSingleSignOnAccount(SingleAccountHelper.java:53)
2019-10-16 17:54:37.396 5945-5945/it.niedermann.owncloud.notes W/System.err:     at it.niedermann.owncloud.notes.persistence.NoteServerSyncHelper.updateAccount(NoteServerSyncHelper.java:121)
2019-10-16 17:54:37.396 5945-5945/it.niedermann.owncloud.notes W/System.err:     at it.niedermann.owncloud.notes.persistence.NoteServerSyncHelper.<init>(NoteServerSyncHelper.java:103)
2019-10-16 17:54:37.396 5945-5945/it.niedermann.owncloud.notes W/System.err:     at it.niedermann.owncloud.notes.persistence.NoteServerSyncHelper.getInstance(NoteServerSyncHelper.java:55)
2019-10-16 17:54:37.396 5945-5945/it.niedermann.owncloud.notes W/System.err:     at it.niedermann.owncloud.notes.persistence.NoteSQLiteOpenHelper.<init>(NoteSQLiteOpenHelper.java:82)
2019-10-16 17:54:37.396 5945-5945/it.niedermann.owncloud.notes W/System.err:     at it.niedermann.owncloud.notes.persistence.NoteSQLiteOpenHelper.getInstance(NoteSQLiteOpenHelper.java:87)
2019-10-16 17:54:37.396 5945-5945/it.niedermann.owncloud.notes W/System.err:     at it.niedermann.owncloud.notes.android.activity.NotesListViewActivity.onCreate(NotesListViewActivity.java:203)
2019-10-16 17:54:37.396 5945-5945/it.niedermann.owncloud.notes W/System.err:     at android.app.Activity.performCreate(Activity.java:6679)
2019-10-16 17:54:37.396 5945-5945/it.niedermann.owncloud.notes W/System.err:     at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1118)
2019-10-16 17:54:37.396 5945-5945/it.niedermann.owncloud.notes W/System.err:     at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2618)
2019-10-16 17:54:37.396 5945-5945/it.niedermann.owncloud.notes W/System.err:     at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2726)
2019-10-16 17:54:37.396 5945-5945/it.niedermann.owncloud.notes W/System.err:     at android.app.ActivityThread.-wrap12(ActivityThread.java)
2019-10-16 17:54:37.396 5945-5945/it.niedermann.owncloud.notes W/System.err:     at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1477)
2019-10-16 17:54:37.396 5945-5945/it.niedermann.owncloud.notes W/System.err:     at android.os.Handler.dispatchMessage(Handler.java:102)
2019-10-16 17:54:37.396 5945-5945/it.niedermann.owncloud.notes W/System.err:     at android.os.Looper.loop(Looper.java:154)
2019-10-16 17:54:37.396 5945-5945/it.niedermann.owncloud.notes W/System.err:     at android.app.ActivityThread.main(ActivityThread.java:6119)
2019-10-16 17:54:37.396 5945-5945/it.niedermann.owncloud.notes W/System.err:     at java.lang.reflect.Method.invoke(Native Method)
2019-10-16 17:54:37.398 5945-5945/it.niedermann.owncloud.notes W/System.err:     at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:886)
2019-10-16 17:54:37.398 5945-5945/it.niedermann.owncloud.notes W/System.err:     at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:776)
2019-10-16 17:54:37.398 5945-5945/it.niedermann.owncloud.notes V/NoteServerSyncHelper: Reinstanziation NotesClient because of SSO acc changed
2019-10-16 17:54:37.408 5945-5945/it.niedermann.owncloud.notes D/NoteServerSyncHelper: Network connection established.
2019-10-16 17:54:37.411 5945-5945/it.niedermann.owncloud.notes V/DatabaseIndexUtil: Creating database index: CREATE INDEX IF NOT EXISTS ACCOUNTS_URL_idx ON ACCOUNTS(URL)
2019-10-16 17:54:37.411 5945-5945/it.niedermann.owncloud.notes V/DatabaseIndexUtil: Creating database index: CREATE INDEX IF NOT EXISTS ACCOUNTS_USERNAME_idx ON ACCOUNTS(USERNAME)
2019-10-16 17:54:37.411 5945-5945/it.niedermann.owncloud.notes V/DatabaseIndexUtil: Creating database index: CREATE INDEX IF NOT EXISTS ACCOUNTS_ACCOUNT_NAME_idx ON ACCOUNTS(ACCOUNT_NAME)
2019-10-16 17:54:37.411 5945-5945/it.niedermann.owncloud.notes V/DatabaseIndexUtil: Creating database index: CREATE INDEX IF NOT EXISTS ACCOUNTS_ETAG_idx ON ACCOUNTS(ETAG)
2019-10-16 17:54:37.411 5945-5945/it.niedermann.owncloud.notes V/DatabaseIndexUtil: Creating database index: CREATE INDEX IF NOT EXISTS ACCOUNTS_MODIFIED_idx ON ACCOUNTS(MODIFIED)
2019-10-16 17:54:37.412 5945-5945/it.niedermann.owncloud.notes V/DatabaseIndexUtil: Creating database index: CREATE INDEX IF NOT EXISTS NOTES_ACCOUNT_ID_idx ON NOTES(ACCOUNT_ID)
2019-10-16 17:54:37.415 5945-5945/it.niedermann.owncloud.notes V/DatabaseIndexUtil: Creating database index: CREATE INDEX IF NOT EXISTS NOTES_REMOTEID_idx ON NOTES(REMOTEID)
2019-10-16 17:54:37.415 5945-5945/it.niedermann.owncloud.notes V/DatabaseIndexUtil: Creating database index: CREATE INDEX IF NOT EXISTS NOTES_ACCOUNT_ID_idx ON NOTES(ACCOUNT_ID)
2019-10-16 17:54:37.415 5945-5945/it.niedermann.owncloud.notes V/DatabaseIndexUtil: Creating database index: CREATE INDEX IF NOT EXISTS NOTES_STATUS_idx ON NOTES(STATUS)
2019-10-16 17:54:37.416 5945-5945/it.niedermann.owncloud.notes V/DatabaseIndexUtil: Creating database index: CREATE INDEX IF NOT EXISTS NOTES_FAVORITE_idx ON NOTES(FAVORITE)
2019-10-16 17:54:37.416 5945-5945/it.niedermann.owncloud.notes V/DatabaseIndexUtil: Creating database index: CREATE INDEX IF NOT EXISTS NOTES_CATEGORY_idx ON NOTES(CATEGORY)
2019-10-16 17:54:37.417 5945-5945/it.niedermann.owncloud.notes V/DatabaseIndexUtil: Creating database index: CREATE INDEX IF NOT EXISTS NOTES_MODIFIED_idx ON NOTES(MODIFIED)
2019-10-16 17:54:37.443 5945-5945/it.niedermann.owncloud.notes W/Glide: Failed to find GeneratedAppGlideModule. You should include an annotationProcessor compile dependency on com.github.bumptech.glide:compiler in your application and a @GlideModule annotated AppGlideModule implementation or LibraryGlideModules will be silently ignored
2019-10-16 17:54:37.509 5945-5982/it.niedermann.owncloud.notes D/NetworkSecurityConfig: Using Network Security Config from resource network_security_config debugBuild: true
2019-10-16 17:54:37.517 5945-5945/it.niedermann.owncloud.notes W/NotesListViewActivity: Tried to update username in drawer, but localAccount was null
2019-10-16 17:54:37.528 5945-5945/it.niedermann.owncloud.notes D/SSOUtil: Permission granted.
2019-10-16 17:54:37.825 5945-5945/it.niedermann.owncloud.notes D/NoteServerSyncHelper: Network connection established.
2019-10-16 17:54:37.931 5945-6005/it.niedermann.owncloud.notes I/OpenGLRenderer: Initialized EGL, version 1.4
2019-10-16 17:54:37.931 5945-6005/it.niedermann.owncloud.notes D/OpenGLRenderer: Swap behavior 1
2019-10-16 17:54:37.934 5945-6005/it.niedermann.owncloud.notes W/OpenGLRenderer: Failed to choose config with EGL_SWAP_BEHAVIOR_PRESERVED, retrying without...
2019-10-16 17:54:37.934 5945-6005/it.niedermann.owncloud.notes D/OpenGLRenderer: Swap behavior 0
2019-10-16 17:54:37.945 5945-6005/it.niedermann.owncloud.notes D/EGL_emulation: eglCreateContext: 0x74f5e8fdcba0: maj 2 min 0 rcv 2
2019-10-16 17:54:37.992 5945-6005/it.niedermann.owncloud.notes D/EGL_emulation: eglMakeCurrent: 0x74f5e8fdcba0: ver 2 0 (tinfo 0x74f5e8ee3f80)
2019-10-16 17:54:38.011 5945-6005/it.niedermann.owncloud.notes D/EGL_emulation: eglMakeCurrent: 0x74f5e8fdcba0: ver 2 0 (tinfo 0x74f5e8ee3f80)
2019-10-16 17:54:38.190 5945-5958/it.niedermann.owncloud.notes V/NoteSQLiteOpenHelper: ID = ? AND STATUS != ? AND ACCOUNT_ID = ?    ----   64 LOCAL_DELETED -1
2019-10-16 17:54:38.190 5945-5958/it.niedermann.owncloud.notes E/SingleNoteWidget: Error: note not found
2019-10-16 17:54:38.201 5945-5957/it.niedermann.owncloud.notes V/NoteSQLiteOpenHelper: ID = ? AND STATUS != ? AND ACCOUNT_ID = ?    ----   64 LOCAL_DELETED -1
2019-10-16 17:54:38.201 5945-5957/it.niedermann.owncloud.notes E/SingleNoteWidget: Error: note not found
2019-10-16 17:54:38.449 5945-6005/it.niedermann.owncloud.notes D/EGL_emulation: eglMakeCurrent: 0x74f5e8fdcba0: ver 2 0 (tinfo 0x74f5e8ee3f80)
2019-10-16 17:54:39.467 5945-6005/it.niedermann.owncloud.notes D/EGL_emulation: eglMakeCurrent: 0x74f5e8fdcba0: ver 2 0 (tinfo 0x74f5e8ee3f80)
2019-10-16 17:54:48.137 5945-5945/it.niedermann.owncloud.notes D/SSOUtil: Permission granted.
2019-10-16 17:54:48.396 5945-6005/it.niedermann.owncloud.notes D/EGL_emulation: eglMakeCurrent: 0x74f5e8fdcba0: ver 2 0 (tinfo 0x74f5e8ee3f80)
2019-10-16 17:54:48.400 5945-6005/it.niedermann.owncloud.notes D/EGL_emulation: eglMakeCurrent: 0x74f5e8fdcba0: ver 2 0 (tinfo 0x74f5e8ee3f80)
2019-10-16 17:54:51.750 5945-6005/it.niedermann.owncloud.notes D/EGL_emulation: eglMakeCurrent: 0x74f5e8fdcba0: ver 2 0 (tinfo 0x74f5e8ee3f80)
2019-10-16 17:54:53.928 5945-5945/it.niedermann.owncloud.notes W/NotesListViewActivity: Tried to update username in drawer, but localAccount was null
2019-10-16 17:54:53.931 5945-5945/it.niedermann.owncloud.notes D/SSOUtil: Permission granted.
2019-10-16 17:54:53.994 5945-5950/it.niedermann.owncloud.notes I/art: Do partial code cache collection, code=21KB, data=29KB
2019-10-16 17:54:53.994 5945-5950/it.niedermann.owncloud.notes I/art: After code cache collection, code=20KB, data=29KB
2019-10-16 17:54:53.994 5945-5950/it.niedermann.owncloud.notes I/art: Increasing code cache capacity to 128KB
2019-10-16 17:54:54.085 5945-6005/it.niedermann.owncloud.notes D/EGL_emulation: eglMakeCurrent: 0x74f5e8fdcba0: ver 2 0 (tinfo 0x74f5e8ee3f80)
2019-10-16 17:54:54.416 5945-6005/it.niedermann.owncloud.notes D/EGL_emulation: eglMakeCurrent: 0x74f5e8fdcba0: ver 2 0 (tinfo 0x74f5e8ee3f80)
2019-10-16 17:54:55.565 5945-6005/it.niedermann.owncloud.notes D/EGL_emulation: eglMakeCurrent: 0x74f5e8fdcba0: ver 2 0 (tinfo 0x74f5e8ee3f80)
```

**Comment by stefan-niedermann** [comment]
Ah, this case

**Comment by dan0xii** [comment]
Oh dear. :)
I will take a look as well but for now I've pushed the appwidget changes because _I think_ it should work once we've ironed out the note id problem.

**Comment by stefan-niedermann** [comment]
@dan0xii i think i fixed the note id bug (stupid me) - Might you double-check if this issue is solved? At least on my devices, i do not have duplicate notes (with different IDs) and the widgets were migrated as expected.

**Comment by stefan-niedermann** [comment]
Oh, i forgot to mention @dan0xii and @korelstar :

Since we now depend in changes in the files app which have not yet been released (for server headers) you will need to checkout files app on the `ssoHeaders`-branch and install this on your virtual devices, at least until these changes get merged into master.

**Comment by dan0xii** [comment]
With the ssoHeaders-branch (logged in) and v1.0.0 of the notes branch, when I upgrade to the multi-account branch I get a prompt to allow Notes to access my contacts but then I'm dropped at a blank note list activity.

I can go to the drawer and I'll already see my nextcloud account as though it is logged in to Notes but when I click on it I get the "Choose an account" dialog. All works as expected after that.

**Comment by stefan-niedermann** [comment]
> With the ssoHeaders-branch (logged in) and v1.0.0 of the notes branch, when I upgrade to the multi-account branch I get a prompt to allow Notes to access my contacts but then I'm dropped at a blank note list activity.

I can go to the drawer and I'll already see my nextcloud account as though it is logged in to Notes but when I click on it I get the "Choose an account" dialog. All works as expected after that.

Should be fixed with the latest commit.

**Comment by korelstar** [comment]
Fixed lastmodified and ETag. Seems to work fine with Nextcloud Android App 3.9.0 from store (no dev version required anymore).
But I did not a code review :wink:

**Comment by stefan-niedermann** [comment]
Okay @dan0xii @korelstar 

i played around and tested various device configurations and it seems to work as far as i can say.

My proposal is to merge this branch into master and release it then, to get some feedback from the F-Droid community.

What do you think? Any major points left, which should be handled before a merge? Any showstopper from you?

**Comment by korelstar** [comment]
I just updated another installation and had an issue with migrating the data: all notes have `modified=null` in the local database. When I then update a note (local or remote), `modified` is updated correctly for that note, but all other notes are shown under `1970` and still have `modified=null`. I have to clear the local app data to enforce a complete resynchronization.

I think this should be fixed before doing a release. Unfortunately, currently I have no time to debug this more. But maybe this happen also to you, when migrating from an old release to the latest version?

**Comment by stefan-niedermann** [comment]
Indeed, i can reproduce this.

On upgrade the following request is being executed:

`2019-12-14 12:09:08.733 21677-21760/it.niedermann.owncloud.notes V/NotesClient: NextcloudRequest: NextcloudRequest(method=GET, header={If-None-Match=["414f4cd580732c5a8a8d8fbe446781ae"]}, parameter={pruneBefore=1576321747}, requestBody=null, url=/index.php/apps/notes/api/v0.2/notes, token=xxxxxxxx, packageName=null, accountName=xxxx@xxxx, followRedirects=false)`

The interesting part is the response here, which only contains a list of the remote ids:

```
...

2019-12-14 12:09:08.757 21677-21760/it.niedermann.owncloud.notes V/---: {"id":428}
2019-12-14 12:09:08.757 21677-21760/it.niedermann.owncloud.notes V/---: {"id":455}
2019-12-14 12:09:08.758 21677-21760/it.niedermann.owncloud.notes V/---: {"id":38306}
2019-12-14 12:09:08.758 21677-21760/it.niedermann.owncloud.notes V/---: {"id":39338}
2019-12-14 12:09:08.758 21677-21760/it.niedermann.owncloud.notes V/---: {"id":45835}
2019-12-14 12:09:08.758 21677-21760/it.niedermann.owncloud.notes V/---: {"id":469}
...
```

Compared to an "old" request, where the json contained the fileds `etag`, `modified`, `title` and so on.

I am wondering where this comes from, because the header `exclude` is nowhere set (i thought this might be an explanation for the missing fields).

**Comment by korelstar** [comment]
I assume that this is fine. Due to `pruneBefore`, all notes before that date are reduced to the id. But I have an idea and will provide a commit, soon.

**Comment by korelstar** [comment]
Please test again with this change (I assume it was a copy/paste error)

**Comment by stefan-niedermann** [comment]
Aye, this solves the problem :) great!

**Comment by korelstar** [comment]
Then LGTM!

**Comment by stefan-niedermann** [comment]
So, thank you all very very much, let's continue the polishing in `master` :)

**Pull request merged**